### PR TITLE
Cupy zonal

### PR DESCRIPTION
**Pull request by kiliakis**
Implements in `cupy` the `zonal.stats` method

## Proposed Changes

  - Implementation of zonal stats in cupy
  - supports both the default stat functions and user defined, custom functions
  - adds asv benchmarks, can be used as a usage example and to demonstrate the speedup w.r.t. the numpy implementation.

### CODE REVIEW

File: benchmarks/benchmarks/zonal.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        W = H = raster_dim
+        zW = zH = zone_dim
+        # Make sure that the raster dim is multiple of the zones dim
+        assert(W % zW == 0)
+        assert(H % zH == 0)
+        # initialize the values raster
+        self.values = get_xr_dataarray((H, W), backend)
+
+        # initialize the zones raster
+        zones = xr.DataArray(np.zeros(H * W).reshape(H, W))
```
Comment (ianthomas23) [inline benchmarks/benchmarks/zonal.py:42]: This can be
```
zones = xr.DataArray(np.zeros((H, W)))
```

File: benchmarks/benchmarks/zonal.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                    acc += x * x
+                return np.sqrt(acc)
+
+            self.custom_stats = {
+                'double_sum': lambda val: val.sum()*2,
+                'l2norm': lambda val: np.sqrt(np.sum(val * val)),
+                'l2normKernel': lambda val: l2normKernel(val)
+            }
+
+    def time_stats_default(self, raster_dim, zone_dim, backend):
```
Comment (ianthomas23) [inline benchmarks/benchmarks/zonal.py:95]: Can you put `zonal` in the benchmark name, for consistency with the other benchmarks. Perhaps `time_zonal_stats_default`?

File: benchmarks/benchmarks/zonal.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        zonal.stats(zones=self.zones, values=self.values)
+
+    def time_stats_custom(self, raster_dim, zone_dim, backend):
```
Comment (ianthomas23) [inline benchmarks/benchmarks/zonal.py:98]: As above.

File: xrspatial/zonal.py [hunk 4]
```diff
@@ -282,6 +296,81 @@ def _stats_numpy(
...
+
+    # stats columns
+    stats_dict = {'zone': []}
+    for stats in stats_funcs:
+        stats_dict[stats] = []
+
+    for i in range(len(unique_zones)):
+        zone_id = unique_zones[i]
+        # skip zone_id == nodata_zones, and non-finite zone ids
+        if (not np.isfinite(zone_id)):
```
Comment (ianthomas23) [inline xrspatial/zonal.py:348]: Don't need the outer brackets.

**Review by ianthomas23 — CHANGES_REQUESTED** [review]
There are just a few small changes required.

The big item missing is a test. But there is another PR open that improves the testing framework so there is no point in adding a test until that is merged. I think the best plan is to merge this first, then I will deal with merge conflicts on the other PR, then we can add a test for this after that.

File: benchmarks/benchmarks/zonal.py (hunk 1 above)
Comment (kiliakis) [reply benchmarks/benchmarks/zonal.py]: Good point!

File: benchmarks/benchmarks/zonal.py (hunk 2 above)
Comment (kiliakis) [reply benchmarks/benchmarks/zonal.py]: Done!

File: benchmarks/benchmarks/zonal.py (hunk 3 above)
Comment (kiliakis) [reply benchmarks/benchmarks/zonal.py]: Done!

File: xrspatial/zonal.py [hunk 5]
```diff
@@ -244,6 +256,8 @@ def _stats_dask_numpy(
 
     return stats_df
 
+# @timing.timeit(key="stats_numpy")
```
[the remark this answers is not in the document]
Comment (kiliakis) [reply xrspatial/zonal.py]: Ooops, removed it

File: xrspatial/zonal.py (hunk 4 above)
Comment (kiliakis) [reply xrspatial/zonal.py]: Yep, outer brackets gone

**Comment by kiliakis** [comment]
Alright cool, latest commit fixed all these minor issues. Will wait for the other PR that deals with the testing infrastructure to add unit-tests for cupy zonal stats!

**Review by ianthomas23 — APPROVED** [review]

**Pull request merged**